File: lib/solutions/day18.ex
defmodule Solutions.Day18 do
  defp init_grid(rows, cols, fallen_bytes) do
    Enum.map(0..rows, fn r ->
      Enum.map(0..cols, fn c ->
        if MapSet.member?(fallen_bytes, {r, c}) do
          "#"
        else
          "."
        end
      end)
    end)
  end

  defp is_valid_block!(grid, r, c) do
    len = length(grid)
    wid = length(hd(grid))
    r < 0 or r >= len or c < 0 or c >= wid or grid |> Enum.at(r) |> Enum.at(c) != "#"
  end

  defp dfs(grid, r, c, seen, fallen_bytes, end_r, end_c, dist) do
    if r == end_r && c == end_c do
      dist
    else
      seen = MapSet.put(seen, {r, c})

      valid_dirs =
        [
          {1, 0},
          {-1, 0},
          {0, 1},
          {0, -1}
        ]
        |> Enum.filter(fn {dr, dc} ->
          is_valid_block!(grid, r + dr, c + dc) &&
          not MapSet.member?(seen, {r + dr, c + dc})
        end)

      valid_dirs
      |> Enum.map(fn {dr, dc} ->
        dfs(grid, r + dr, c + dc, seen, fallen_bytes, end_r, end_c, dist + 1)
      end)
    end
  end

  def solve(lines) do
    # for test case: {rows, cols} = {6, 6}
    {rows, cols} = {6, 6}

    fallen_bytes =
      lines
      |> Enum.reduce_while(MapSet.new(), fn coord, blocks ->
        if MapSet.size(blocks) == 12 do
          {:halt, blocks}
        else
          [col, row] =
            coord
            |> String.split(",")
            |> Enum.map(&String.to_integer(&1))

          {:cont, MapSet.put(blocks, {row, col})}
        end
      end)
      |> IO.inspect()

    memory = init_grid(rows, cols, fallen_bytes)

    part1 =
      dfs(memory, 0, 0, MapSet.new(), fallen_bytes, rows, cols, 0)
      |> IO.inspect()

    {0, 0}
  end
end

File: lib/solutions/day18.py
from collections import deque

def bfs(rows, cols, fallen_bytes, limit):
    memory = [["." for _ in range(cols + 1)] for _ in range(rows + 1)]
    for col, row in fallen_bytes[:limit]:
        memory[row][col] = "#"
        
    queue = deque([(0, 0, 0)])
    seen = {(0, 0)}

    while queue:
        r, c, dist = queue.popleft()            
        if r == rows and c == cols:
            return (True, dist)
        for nr, nc in [(r + 1, c), (r - 1, c), (r, c + 1), (r, c - 1)]:
            if nr < 0 or nr > rows or nc < 0 or nc > cols or memory[nr][nc] == "#":
                continue
            if (nr, nc) in seen:
                continue        
            seen.add((nr, nc))
            queue.append((nr, nc, dist + 1))

    return (False, 0)

with open("lib/inputs/day18.txt") as f:
    # rows, cols = 6, 6
    # limit = 12
    rows, cols = 70, 70
    limit = 1024
    fallen_bytes = [list(map(int, cord.split(","))) for cord in f.read().strip().split("\n")]
        
    part1 = bfs(rows, cols, fallen_bytes, limit)
    print(part1)

    part2 = (0,0)
    for i in range(limit, len(fallen_bytes)):
        if not bfs(rows, cols, fallen_bytes, i)[0]:
            part2 = fallen_bytes[i-1]
            break
    print(part2)

